### ISSUE TITLE
Possible Crash on ReadWrite.write(after:_)

### ISSUE THREAD

**Issue by Enricoza** [Tealium/tealium-swift#112]
**Your Environment**
Tealium Swift library version: 1.8.0-beta2

**Describe the bug**
Possible crashes can happen because of writes that are made concurrently on _ReadWrite.write(after:_)_ method.
This is due to the fact that the _barrier_ flag is missing on the _queue.asyncAfter_ method and the _self.write(work)_ method will always execute the work block without other dispatches due to _isAlreadyInQueue_ returning true in this case.

**To Reproduce**
To reproduce a crash you can just execute this block of code:

```
        let readWrite = ReadWrite("TEST")
        var aString = ""
        DispatchQueue.concurrentPerform(iterations: 10000) { (i) in
            readWrite.write(after: DispatchTime.now() + .seconds(1)) {
                aString += "\(i)"
            }
        }
```

**Expected behavior**
Writes should never be concurrent.

**Possible Solutions**
I think it's safe to add the _barrier_ flag on the _dispatchAsync_ method (it will only apply the barrier while actually executing the block inside) and execute like this:

```
    public func write(after delay: DispatchTime, _ work: @escaping () -> Void) {
        queue.asyncAfter(deadline: delay, flags: .barrier) {
            work()
        }
    }
```

Otherwise (but I wouldn't recommend it) you can either firstly dispatch on another queue: 
```
    let anotherQueue = DispatchQueue.global() // Or any other
    public func write(after delay: DispatchTime, _ work: @escaping () -> Void) {
        anotherQueue.asyncAfter(deadline: delay, flags: .barrier) {
            self.write(work)
        }
    }
```
or (which also seems bad) directly dispatch it again on the same queue:

```
    public func write(after delay: DispatchTime, _ work: @escaping () -> Void) {
        queue.asyncAfter(deadline: delay) {
            queue.async(flags: .barrier, execute: work)
        }
    }
```

Hope this helps

PS: I didn't actually come across some crashes (I didn't go to production with the beta yet) so I'm not 100% sure that this would actually cause many crashes, but I think this could be the case (if this method is actually called a considerable amount of times inside the sdk)

**Comment by craigrouse** [comment]
@Enricoza thanks for taking the time to look at this. We'll have a look and get back to you.

**Comment by Enricoza** [comment]
@craigrouse since I was making some tests which turned to be successful, I'm opening a PR with a solution to this problem and some other (possible) problems with nested reads and writes

**Comment by craigrouse** [comment]
Thanks @Enricoza! I tested your fix, and was in the process of putting together a new release with it included, but will take a look at your PR. Appreciate your efforts, as ever.

**Comment by craigrouse** [comment]
Resolved in 1.8.0 release. Thanks for your help @Enricoza!
